### PR TITLE
rfc77_drop_python2_support - RFC 77: Drop Python 2 support in favor of Python 3.6

### PR DESCRIPTION
**Pull request by idanmiara**
## What does this PR do?

Add RFC 77: Drop Python 2 support in favor of Python 3.6

## What are related issues/pull requests?

https://github.com/OSGeo/gdal/issues/3114

## Tasklist

 - [ ] ADD YOUR TASKS HERE
 - [ ] Add test case(s)
 - [ ] Add documentation
 - [x] Review
 - [x] Adjust for comments
 - [x] All CI builds and checks have passed

### CODE REVIEW

**Comment by idanmiara** [comment]
https://github.com/talos-gis/gdal/blob/Branch_rfc_drop_python2/gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst

**Comment by jratike80** [comment]
I think that sentense "Keeping Python 2.7 support virtually serves no purpose as virtually all major OS and relevant programs already dropped Python 2 support" oversimplifies the situation. There must be loads of existing Python stuff that is written for Python 2 and that is not just GDAL. Individuals and and bigger and smaller companies who are using such code may be forced to patch their code and meanwhile to run both Python 2 and Python 3.

I guess that upgrading is something that users can't avoid anyway for long. I just wanted to express that keeping Python 2.7 support _might_ serve _some_ purpose.

**Comment by rcoup** [comment]
> in favor of Python 3.6|3.7

I think the RFC should pick one, not leave a decision open to further discussion.

File: gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+Currently GDAL Python bindings support Python 2.7 and Python 3.x (x=?).
+Python 2 is at End Of Life, and is no longer supported since the January 2020.
+Keeping Python 2.7 support virtually serves no purpose as virtually all major OS
+and relevant programs already dropped Python 2 support.
+On the other hand, keeping Python 2 support puts unnecessary maintenance burden on the GDAL maintainers,
+As maintainers need to make sure their new code is backwards compatible with Python 2.
+Moreover, many important features that added in Python 3 cannot be used in GDAL to maintain backwards compatibility with Python 2.
+
+Incomplete features list of Python 3.6, several of which are immediately useful to simplify code or improve testing:
```
Comment (rcoup) [inline gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst:32]: this is only features of 3.6 cf 3.5... not 3.x over 2.7 - of which there are many more.

Comment (idanmiara) [reply gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst:32]: Right, Do you want to list any specific ones that worth noting ?

**Comment by idanmiara** [comment]
> I think that sentense "Keeping Python 2.7 support virtually serves no purpose as virtually all major OS and relevant programs already dropped Python 2 support" oversimplifies the situation. There must be loads of existing Python stuff that is written for Python 2 and that is not just GDAL. Individuals and and bigger and smaller companies who are using such code may be forced to patch their code and meanwhile to run both Python 2 and Python 3.

Good point, I will add a note about that.

> Perhaps RFC could include some kind of SLA to inform users how long GDAL 3.2.x is maintained and receives at least security fixes?

There was a (non conclusive?) discussion about making an LTS version when GDAL 3.1.0 was released:
http://osgeo-org.1560.x6.nabble.com/gdal-dev-Reconsidering-release-cycle-length-td5436163.html#a5436242
maybe @rouault  can elaborate on this matter.

**Comment by idanmiara** [comment]
> > in favor of Python 3.6|3.7
> 
> I think the RFC should pick one, not leave a decision open to further discussion.

I was thinking that 3.6 could be indeed a better choice for this first step. 
What do you think about future regular dropping AKA NEP 29?

**Comment by rouault** [comment]
> There was a (non conclusive?) discussion about making an LTS version when GDAL 3.1.0 was released:
> http://osgeo-org.1560.x6.nabble.com/gdal-dev-Reconsidering-release-cycle-length-td5436163.html#a5436242
> maybe @rouault can elaborate on this matter.

Nobody has stepped up to make a LTS, so there won't be one. We only provide bugfix releases of the current stable branch for 6 months.

**Comment by idanmiara** [comment]
I think I've finished my research for this RFC...
Any comments ?

**Comment by idanmiara** [comment]
I can't figure out why Make html fails.

**Comment by rouault** [comment]
> I can't figure out why Make html fails.

From the logs:
```
/__w/1/s/gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst:223: WARNING: Document or section may not begin with a transition.
/__w/1/s/gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst:228: WARNING: Document or section may not begin with a transition.
/__w/1/s/gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst:228: WARNING: Document may not end with a transition.
```

**Comment by idanmiara** [comment]
> /__w/1/s/gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst:223: WARNING: Document or section may not begin with a transition.

Thanks!

**Comment by idanmiara** [comment]
What is the process for publishing a RFC?
Do I need to do anything else?

File: gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+Backward compatibility
+----------------------
+
+Currently, GDAL Python code itself is compatible with Python 2 and Python 3.
+Once this PR is accepted, GDAL 3.3.0 would not be compatible with Python 2.
+Thus any "Python 2 only" code that uses GDAL would need to be upgraded to Python 3 and
+at the same time the respective Python interpreter would need to be upgraded
+to a supported Python version.
+
+Making GDAL 3.2.0 as a LTS version:
```
Comment (rouault) [inline gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst:157]: should be rephrased. On a quick read, one might understand that 3.2.0 will be a LTS version

**Comment by rouault** [comment]
I'd suggest you send an email to gdal-dev entitled "Call for discussion on RFC77 Drop Python 2 support"  and link to that pull request and https://github.com/talos-gis/gdal/blob/Branch_rfc_drop_python2/gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst for nicer display.
We'll merge this PR once we've had some feedback.

I was just thinking that we could potentially emit a deprecation warning in GDAL 3.2.1 if people run Python < 3.6 ?

File: gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst (hunk 2 above)
Comment (idanmiara) [reply gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst:157]: right. done.

**Comment by idanmiara** [comment]
> I'd suggest you send an email to gdal-dev entitled "Call for discussion on RFC77 Drop Python 2 support" and link to that pull request and https://github.com/talos-gis/gdal/blob/Branch_rfc_drop_python2/gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst for nicer display.
> We'll merge this PR once we've had some feedback.

OK, thanks.

> I was just thinking that we could potentially emit a deprecation warning in GDAL 3.2.1 if people run Python < 3.6 ?

I'll try to workout a generalized idea for that.

**Comment by idanmiara** [comment]
> > I'd suggest you send an email to gdal-dev entitled "Call for discussion on RFC77 Drop Python 2 support" and link to that pull request and https://github.com/talos-gis/gdal/blob/Branch_rfc_drop_python2/gdal/doc/source/development/rfc/rfc77_drop_python2_support.rst for nicer display.
> > We'll merge this PR once we've had some feedback.
> 
> OK, thanks.
> 
> > I was just thinking that we could potentially emit a deprecation warning in GDAL 3.2.1 if people run Python < 3.6 ?
> 
> #3165

So far 100% (silent) support from the mailing list for this RFC... :)

**Comment by snowman2** [comment]
Related:
- https://github.com/pyproj4/pyproj/issues/331
- https://github.com/pyproj4/pyproj/issues/499

**Comment by DanielFEvans** [comment]
> So far 100% (silent) support from the mailing list for this RFC... :)

I'll confirm that at least my own silence on the mailing list correlates to support!

Given the overhead of upgrading to GDAL 3, I suspect that any project still stuck on Python 2 will also be stuck on GDAL 2 (or even earlier). That's the case for our last few remaining pieces of software - as and when they're upgraded to GDAL 3.3, they'll be converted to Python 3 at the same time.

**Comment by schwehr** [comment]
Thank you for writing up this RFC.

Python 2.7 support is a drain on everything.  If someone wants to maintain it, they can easily fork.  While I know there are a lot of things like LTS distributions that will be around for long time that are heavily focused on the old world, that isn't the responsibility of the GDAL project.

**Comment by idanmiara** [comment]
@snowman2 @schwehr @DanielFEvans @hobu Thanks for your comments!

**Comment by mmomtchev** [comment]
Ubuntu switch to Python 3 by default in 18.04 and Debian did so only in buster  - the very latest version.
Python 3 has been around for a very long time, but it is only now that the major distributions started switching their default interpreters
Thus said, people who are unable / unwilling to upgrade to Python 3 probably won't upgrade GDAL either 😄 

I would also like to bring this page to your attention:
<https://python3statement.org/>
Maybe you can add GDAL if this RFC passes

**Comment by rcoup** [comment]
@mmomtchev 

> Ubuntu switched to Python 3 by default in 18.04 and Debian did so only in buster - the very latest version.
> Python 3 has been around for a very long time, but it is only now that the major distributions started switching their default interpreters

Python3 has been installable in both distributions for many releases over many years though. Distributions were discouraged from making `python` point to `python3` by the original [PEP-394](https://www.python.org/dev/peps/pep-0394/) until Py2 was basically "gone". The RFC text has info on distribution releases timelines and their associated Python versions.

**Comment by rouault** [comment]
you can update the voting history section with https://lists.osgeo.org/pipermail/gdal-dev/2020-November/053039.html

**Comment by idanmiara** [comment]
Thanks!
Done.

**Pull request merged**